pkg/controller/logstorage/logstorage.go

```diff
@@ -0,0 +1,288 @@
+package logstorage
+
+import (
+	"context"
+	"fmt"
+
+	kbv1 "github.com/elastic/cloud-on-k8s/pkg/apis/kibana/v1"
+	apps "k8s.io/api/apps/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	cmnv1 "github.com/elastic/cloud-on-k8s/pkg/apis/common/v1"
+	esv1 "github.com/elastic/cloud-on-k8s/pkg/apis/elasticsearch/v1"
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	storagev1 "k8s.io/api/storage/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	operatorv1 "github.com/tigera/operator/api/v1"
+	"github.com/tigera/operator/pkg/controller/utils"
+	"github.com/tigera/operator/pkg/controller/utils/imageset"
+	"github.com/tigera/operator/pkg/render"
+	relasticsearch "github.com/tigera/operator/pkg/render/common/elasticsearch"
+	rmeta "github.com/tigera/operator/pkg/render/common/meta"
+	rsecret "github.com/tigera/operator/pkg/render/common/secret"
+)
+
+func (r *ReconcileLogStorage) createLogStorage(
+	ls *operatorv1.LogStorage,
+	install *operatorv1.InstallationSpec,
+	variant operatorv1.ProductVariant,
+	clusterConfig *relasticsearch.ClusterConfig,
+	managementCluster *operatorv1.ManagementCluster,
+	managementClusterConnection *operatorv1.ManagementClusterConnection,
+	esService *corev1.Service,
+	kbService *corev1.Service,
+	pullSecrets []*corev1.Secret,
+	hdler utils.ComponentHandler,
+	reqLogger logr.Logger,
+	ctx context.Context,
+) error {
+	var esInternalCertSecret, esAdminUserSecret, esCertSecret *corev1.Secret
+	var kibanaSecrets, curatorSecrets []*corev1.Secret
+	var esLicenseType render.ElasticsearchLicenseType
+
+	if managementClusterConnection == nil {
+		// Get the admin user secret to copy to the operator namespace.
+		esAdminUserSecret, err := utils.GetSecret(ctx, r.client, render.ElasticsearchAdminUserSecret, render.ElasticsearchNamespace)
+		if err != nil {
+			reqLogger.Error(err, err.Error())
+			return err
+		}
+		if esAdminUserSecret != nil {
+			esAdminUserSecret = rsecret.CopyToNamespace(rmeta.OperatorNamespace(), esAdminUserSecret)[0]
+		}
+		r.esAdminUserSecret = esAdminUserSecret
+
+		// Check if there is a StorageClass available to run Elasticsearch on.
+		if err := r.client.Get(ctx, client.ObjectKey{Name: ls.Spec.StorageClassName}, &storagev1.StorageClass{}); err != nil {
+			if errors.IsNotFound(err) {
+				err := fmt.Errorf("couldn't find storage class %s, this must be provided", ls.Spec.StorageClassName)
+				reqLogger.Error(err, err.Error())
+				r.status.SetDegraded("Failed to get storage class", err.Error())
+				return err
```

Comment:
I've added comments at the beginning of this file, but wanted to point one out in particular, this cannot return an error or we'll end up in a tight Reconcile loop.